File: gff/write_fasta_from_gff.pl
#!/usr/bin/env perl

=head1 NAME

write_fasta_from_gff.pl - Extracts the protein or CDS sequences from a GFF3 file

=head1 SYNOPSIS

USAGE: write_fasta_from_gff.pl 
            --gff=/path/to/some_file.gff
            --output=/path/to/somefile.faa
          [ --type=protein|cds
            --fasta=/path/to/somefile.fna
          ]

=head1 OPTIONS

B<--gff,-g>
    Input file following the GFF3 specification.

B<--output,-o>
    Output FASTA file to be created.

B<--type,-t>
    Optional.  Should the output records be the CDS nucleotides or translated as
    proteins?  (values are 'cds' or 'protein', default=protein)

B<--fasta,-f>
    Optional.  If the FASTA entries for the underlying assemblies isn't 
    included in the GFF3 document you'll need to pass a FASTA file here.

B<--log,-l> 
    Log file

B<--help,-h>
    This help message

=head1  DESCRIPTION

This extracts the CDS entries from a GFF3 file, optionally translates them, and
exports a FASTA file of the CDS or resulting polypeptide sequences.

=head1  INPUT

The input file should follow the official GFF3 specification found here:

    http://www.sequenceontology.org/gff3.shtml

=head1  OUTPUT

A FASTA file with the CDS entries extracted and optionally translated to polypeptides.

=head1  CONTACT

    Joshua Orvis
    jorvis@gmail.com

=cut

use strict;
use Getopt::Long qw(:config no_ignore_case no_auto_abbrev pass_through);
use Pod::Usage;
use FindBin;
use lib "$FindBin::Bin/../lib";
use bioUtils;

my %options = ();
my $results = GetOptions (\%options, 
                          'gff|g=s',
                          'output|o=s',
                          'fasta|f=s',
                          'type|t=s',
                          'use_locus_tags!',
                          'log|l=s',
                          'help|h') || pod2usage();

## display documentation
if( $options{'help'} ){
    pod2usage( {-exitval => 0, -verbose => 2, -output => \*STDERR} );
}

## make sure everything passed was peachy
&check_parameters(\%options);

## open the log if requested
my $logfh;
if (defined $options{log}) {
    open($logfh, ">$options{log}") || die "can't create log file: $!";
}

## structure like: 
#   $h{$mol_id}{$mRNA_id}, [ { min => N, max => N, strand => ?, phase => N } ];
my %coords = ();

my %molecules = ();

open(my $gffin_fh, $options{gff}) || die "failed to open input gff file: $options{gff}: $!";
open(my $fastaout_fh, ">$options{output}") || die "failed to create FASTA output file: $!";

my $in_fasta = 0;
my $current_seq_id;
my @current_seq_lines = ();

my %mRNA_locus_tags = ();

my %locus_lookup = ();

while ( my $line = <$gffin_fh> ) {
    chomp $line; 
    
    if ( $line =~ /^\#/ ) {
        next;
    }
    
    my @cols = split("\t", $line);
    
    if ( $line =~ /^\#\#FASTA/ ) {
        $in_fasta = 1;
        next;

    } elsif ( $in_fasta ) {
        if ( $line =~ /^>(\S+)/ ) {
            my $next_seq_id = $1;

            if ( $current_seq_id ) {            
                if ( exists $molecules{$current_seq_id} ) {
                    die "found duplicate molecule ($current_seq_id) in FASTA definitions";
                }            

                $molecules{$current_seq_id} = join( '', @current_seq_lines );
                $molecules{$current_seq_id} =~ s/\s//g;
                $molecules{$current_seq_id} = uc( $molecules{$current_seq_id} );
            }
            
            $current_seq_id = $next_seq_id;
            @current_seq_lines = ();
        } else {
            push @current_seq_lines, $line;
        }
    
    } else {
        next unless scalar @cols == 9;
    
        ## check for Parent definitions with others after 
        my $parent;
        my $locus_tag;
        my $feat_id;
        

        if ( $cols[2] eq 'mRNA') {
            if ($options{use_locus_tags}) {
                if ( $cols[8] =~ /locus_tag\=(.+?)\;/ ) {
                    $locus_tag = $1;
                } elsif ( $cols[8] =~ /locus_tag\=(.+)/ ) {
                    $locus_tag = $1;
                } else {
                    die "Failed to find locus tag on mRNA feature line: $line";
                }

                $cols[8] =~ /ID\=(.+?)\;/ || die "Failed to find ID attribute on the following mRNA line: $line";
                $feat_id = $1;

                $locus_lookup{$feat_id} = $locus_tag;
            }
        }

        if ( $cols[2] eq 'CDS' ) {

            if ( $cols[8] =~ /Parent\=(.+?)\;/ ) {
                $parent = $1;
            } elsif ( $cols[8] =~ /Parent\=(.+)/ ) {
                $parent = $1;
            } else {
                die "CDS features require Parent features to be defined";
            }

            push @{$coords{$cols[0]}{$parent}}, { min => $cols[3], max => $cols[4], 
                                        strand => $cols[6], phase => $cols[7] };

        }
    }
}

## do the last one and reset
if ($current_seq_id) {
    $molecules{$current_seq_id} = join( '', @current_seq_lines );
    $molecules{$current_seq_id} =~ s/\s//g;
    $molecules{$current_seq_id} = uc( $molecules{$current_seq_id} );
}

## sanity check that any sequence data was actually found
if (scalar keys %molecules == 0 && ! $options{fasta}) {
    die "ERROR: No sequence data found embedded in GFF3.  Perhaps you forgot to pass the path to a FASTA file via --fasta ? ";
}

$current_seq_id = '';
@current_seq_lines = ();

if ( $options{fasta} ) {
    open(my $fasta_fh, "<$options{fasta}") || die "failed to read FASTA file: $!";
    
    while (my $line = <$fasta_fh>) {
        if ( $line =~ /^>(\S+)/ ) {
            my $next_seq_id = $1;
            
            if ( $current_seq_id ) {
                if ( exists $molecules{$current_seq_id} ) {
                    die "found duplicate molecule ($current_seq_id) in FASTA definitions";
                }

                $molecules{$current_seq_id} = join( '', @current_seq_lines );
                $molecules{$current_seq_id} =~ s/\s//g;
                $molecules{$current_seq_id} = uc( $molecules{$current_seq_id} );
            }
            
            $current_seq_id = $next_seq_id;
            @current_seq_lines = ();
        } else {
            push @current_seq_lines, $line;
        }
    }
}

## do the last one and reset
$molecules{$current_seq_id} = join( '', @current_seq_lines );
$molecules{$current_seq_id} =~ s/\s//g;
$molecules{$current_seq_id} = uc( $molecules{$current_seq_id} );

for my $mol_id ( keys %coords ) {

    ## make sure this molecule exists in FASTA
    if ( ! exists $molecules{$mol_id} ) {
        die "ERROR: molecule ($mol_id) referenced in GFF but not found in FASTA\n";
    }

    for my $mrna_id ( keys %{$coords{$mol_id}} ) {
        my $complete_cds = '';
        my $strand;

        for my $cds ( sort {$$a{min} <=> $$b{min} } @{$coords{$mol_id}{$mrna_id}} ) {
            $complete_cds .= substr($molecules{$mol_id}, 
                                   $$cds{min} - 1,
                                   $$cds{max} - $$cds{min} + 1);
            
            $strand = $$cds{strand};
        }
        
        if ( $strand eq '-' ) {
            $complete_cds = reverse $complete_cds;
            $complete_cds =~ tr/ATGC/TACG/;
        }

        my $display_id = $mrna_id;

        if ($options{use_locus_tags}) {
            $display_id = $locus_lookup{$mrna_id};
        }
        
        if ($options{type} eq 'protein' ) {
            my $protein_seq = translate_sequence( $complete_cds, $mrna_id );
            print $fastaout_fh ">$display_id\n" . characters_per_line($protein_seq,60) . "\n";
        } else {
            print $fastaout_fh ">$display_id\n" . characters_per_line($complete_cds,60) . "\n";
        }
    }
}


exit(0);


sub _log {
    my $msg = shift;

    print $logfh "$msg\n" if $logfh;
}


sub check_parameters {
    my $options = shift;
    
    ## make sure required arguments were passed
    my @required = qw( gff output );
    for my $option ( @required ) {
        unless  ( defined $$options{$option} ) {
            die "--$option is a required option";
        }
    }
    
    if ( defined $$options{type} ) {
        $$options{type} = lc $$options{type};
    
        if ( $$options{type} ne 'cds' && $$options{type} ne 'protein' ) {
            die "option --type must have a value of either cds or protein";
        }
    } else {
        $$options{type} = 'protein';
    }
}

File: gff/write_fasta_from_gff.py
#!/usr/bin/env python3

"""

This utility reads an annotated GFF3 file and writes FASTA of different feature types,
either the CDS or protein sequences (protein seqs are translated from the CDS).  In
either case, the genomic FASTA must be available, either embedded at the end of the
GFF3 or as a separate file (passed via -f).

The exported IDs will be the ID attributes of the mRNA features unless these have
locus_tag values in the 9th column.  Also, if the polypeptides associated with each
mRNA have product_name values, that will also be written to the FASTA header.

If the --check_ends option is passed, each CDS extracted is checked for canonical start
and stop sites.  When this happens, a warning is printed on STDERR like:

   WARN: Non-canonical start codon (AAA) in mRNA G678.mRNA.9856.1

Warnings are also printed if translating to protein and undefined codons are found:

   WARN: Encountered unknown codon during translation: NTA
   WARN: Non-canonical stop codon (CCA) in mRNA G678.mRNA.5532.1


Author:  Joshua Orvis (jorvis AT gmail)
"""

import argparse
import sys

from biocode import utils, gff


def main():
    parser = argparse.ArgumentParser( description='Extracts the protein or CDS seqeunces from a GFF3 file')

    ## output file to be written
    parser.add_argument('-i', '--input_file', type=str, required=True, help='Path to an input GFF3 file to be read' )
    parser.add_argument('-o', '--output_file', type=str, required=False, help='Path to an output FASTA file to be created' )
    parser.add_argument('-t', '--type', type=str, required=False, default='protein', choices=['protein', 'cds'], help='Type of features to export')
    parser.add_argument('-f', '--fasta', type=str, required=False, help='If the FASTA entries for the underlying assemblies is absent from the GFF3 document passed, you will need to specify this option' )
    parser.add_argument('-ft', '--feature_type', type=str, required=False, default='mRNA', choices=['mRNA', 'polypeptide'], help='IDs and coordinates will come from this feature type' )
    parser.add_argument('--check_ends', dest='check_ends', action='store_true')
    parser.add_argument('--check_internal_stops', dest='check_internal_stops', action='store_true')
    parser.set_defaults(check_ends=False, check_internal_stops=False)
    args = parser.parse_args()

    ## output will either be a file or STDOUT
    fout = sys.stdout
    if args.output_file is not None:
        fout = open(args.output_file, 'wt')

    # sanity option check
    if args.check_internal_stops == True and args.type == 'cds':
        raise Exception("Error:  Checking internal stops for CDS features not currently supported.")

    (assemblies, features) = gff.get_gff3_features(args.input_file)

    # only doing the standard codon table for now
    start_codons = ['ATG', 'GTG', 'TTG']
    stop_codons  = ['TAG', 'TAA', 'TGA']

    ## add sequence residues from external FASTA file if the user passed one
    if args.fasta is not None:
        utils.add_assembly_fasta(assemblies, args.fasta)
    
    for assembly_id in assemblies:
        for gene in assemblies[assembly_id].genes():
            if args.feature_type == 'mRNA':
                feats = gene.mRNAs()
            elif args.feature_type == 'polypeptide':
                feats = gene.polypeptides()
            
            for feat in feats:
                ## initial values of id and header to export (can be overridden by available annotation)
                export_id = feat.id
                export_header = None

                ## Add the gene product name if there is one
                if args.feature_type == 'mRNA':
                    for polypeptide in feat.polypeptides():
                        if polypeptide.annotation is not None:
                            if polypeptide.annotation.product_name is not None:
                                export_header = polypeptide.annotation.product_name
                                break

                    coding_seq = feat.get_CDS_residues(for_translation=True)
                    if feat.locus_tag is not None:
                        export_id = feat.locus_tag
                        
                elif args.feature_type == 'polypeptide':
                    export_header = feat.annotation.product_name
                    coding_seq = feat.parent.get_CDS_residues(for_translation=True)
                    if feat.parent.locus_tag is not None:
                        export_id = feat.parent.locus_tag

                if len(coding_seq) > 0:
                    fout.write(">{0}".format(export_id))
                    if export_header is not None:
                        fout.write(" {0}\n".format(export_header))
                    else:
                        fout.write("\n")

                    if args.check_ends == True:
                        # check the starting codon
                        start_codon = coding_seq[0:3].upper()
                        if start_codon not in start_codons:
                            sys.stderr.write("WARN: Non-canonical start codon ({0}) in mRNA {1}\n".format(start_codon, feat.id))

                        stop_codon = coding_seq[-3:].upper()
                        if stop_codon not in stop_codons:
                            sys.stderr.write("WARN: Non-canonical stop codon ({0}) in mRNA {1}\n".format(stop_codon, feat.id))                        

                    if args.type == 'cds':
                        fout.write("{0}\n".format(utils.wrapped_fasta(coding_seq)))
                    else:
                        translated_seq = utils.translate(coding_seq)

                        if args.check_internal_stops == True:
                            internal_stop_count = translated_seq[:-1].count('*')
                            if internal_stop_count > 0:
                                sys.stderr.write("Found {0} internal stops in mRNA {1}\n".format(internal_stop_count, feat.id))

                        fout.write("{0}\n".format(utils.wrapped_fasta(translated_seq)))
                else:
                    print("WARNING: Skipped feature {0} because it had no associated CDS features".format(export_id), file=sys.stderr)

if __name__ == '__main__':
    main()







